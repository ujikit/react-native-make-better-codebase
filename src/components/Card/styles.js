//package import
import { StyleSheet, Platform } from 'react-native';

//local import
import { COLORS } from '../../configs';

export default StyleSheet.create({
  shadow: {
    shadowColor: COLORS.black,
    shadowOffset: {
      width: 0,
      height: 3,
    },
    shadowOpacity: 0.29,
    shadowRadius: 4.65,
    elevation: 7,
  },
  wrapCardBasic: {
    width: null,
    borderRadius: 10,
    padding: 16,
  },
  wrapCardButton: {
    borderRadius: 10,
    overflow: Platform.OS === 'ios' ? 'visible' : 'hidden',
  },
  containerCardButton: {
    borderRadius: 10,
    alignItems: 'flex-start',
    justifyContent: 'flex-start',
  },
});
