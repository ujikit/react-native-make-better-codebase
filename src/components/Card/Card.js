//package import
import React, { memo } from 'react';
import { View } from 'react-native';
import PropTypes from 'prop-types';

//local import
import styles from './styles';
import { COLORS } from '../../configs';
import Button from '../Button';

const Card = ({
  types,
  children,
  color,
  shadow,
  styleShadow,
  styleWrap,
  styleContainer,
  onPress,
  disabled,
  disabledColor,
  width,
  height,
  ...props
}) => {
  let widthWrap = null;
  let widthContainer = null;
  let heightWrap = null;
  let heightContainer = null;
  let styleShadows =
    Object.keys(styleShadow).length !== 0 ? styleShadow : styles.shadow;

  if (width !== null && width !== undefined) {
    if (width.toString().match(/%/g)) {
      widthWrap = width;
      widthContainer = '100%';
    } else {
      widthWrap = width;
      widthContainer = width;
    }
  }

  if (height !== null && height !== undefined) {
    if (height.toString().match(/%/g)) {
      heightWrap = height;
      heightContainer = '100%';
    } else {
      heightWrap = height;
      heightContainer = height;
    }
  }

  switch (types) {
    case 'basic':
      return (
        <View
          style={[
            styles.wrapCardBasic,
            { width: widthWrap, height: heightWrap },
            shadow && styleShadows,
            styleWrap,
            { backgroundColor: color },
          ]}
        >
          {children}
        </View>
      );
    case 'button':
      return (
        <Button
          onPress={onPress}
          color={color}
          styleWrap={[
            styles.wrapCardButton,
            { width: widthContainer, height: heightContainer },
            styleWrap,
            shadow && styleShadows,
          ]}
          styleContainer={[
            styles.containerCardButton,
            { width: widthContainer, height: heightContainer },
            styleContainer,
          ]}
          disabled={disabled}
          disabledColor={disabledColor}
          {...props}
        >
          {children}
        </Button>
      );
  }
};

Card.propTypes = {
  types: PropTypes.oneOf(['basic', 'button']),
  children: PropTypes.node,
  color: PropTypes.string,
  shadow: PropTypes.bool,
  styleShadow: PropTypes.oneOfType([PropTypes.object, PropTypes.any]),
  styleWrap: PropTypes.oneOfType([PropTypes.object, PropTypes.any]),
  styleContainer: PropTypes.oneOfType([PropTypes.object, PropTypes.any]),
  onPress: PropTypes.func,
  disabled: PropTypes.bool,
  disabledColor: PropTypes.string,
  width: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  height: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
};

Card.defaultProps = {
  types: 'basic',
  children: <View />,
  color: COLORS.primaryWhite,
  shadow: true,
  styleShadow: {},
  styleWrap: {},
  styleContainer: {},
  onPress: () => null,
  disabled: false,
  disabledColor: COLORS.black70,
  width: null,
  height: null,
};

export default memo(Card);
