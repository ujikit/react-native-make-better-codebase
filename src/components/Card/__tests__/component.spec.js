import 'react-native';
import React from 'react';
import renderer from 'react-test-renderer';
import Card from '..';

describe('Some component', () => {
  it('renders correctly', () => {
    const tree = renderer.create(<Card />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
