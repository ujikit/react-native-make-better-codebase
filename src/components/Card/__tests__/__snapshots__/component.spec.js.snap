// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Some component renders correctly 1`] = `
<View
  style={
    Array [
      Object {
        "borderRadius": 10,
        "overflow": "hidden",
        "padding": 16,
        "width": null,
      },
      Object {
        "elevation": 7,
        "shadowColor": "#000000",
        "shadowOffset": Object {
          "height": 3,
          "width": 0,
        },
        "shadowOpacity": 0.29,
        "shadowRadius": 4.65,
      },
      Object {},
      Object {
        "backgroundColor": "#FFFFFF",
      },
    ]
  }
>
  <View />
</View>
`;
